### ISSUE TITLE
Enhancement: [explicit-function-return-type] Allow `satisfies` operator when `allowTypedFunctionExpressions` is set

### ISSUE THREAD

**Issue by Isti115** [typescript-eslint/typescript-eslint#7692]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/explicit-function-return-type

### Description

TypeScript 4.9 introduced the [`satisfies` operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator), which makes it possible to add explicit type annotations inline without compromising type safety. (The currently supported `as` operator completely overrides the type, making it really easy to introduce discrepancies between the annotations and the real types by sidestepping the typechecker.)

### Fail

```typescript
const f = () => '';
```

### Pass

```typescript
const f = (() => '') satisfies () => string;
```

### Additional Info

I have a working commit that I could submit as a PR if this issue gets accepted: https://github.com/Isti115/typescript-eslint/commit/a373ca7db1e7e7c6b91e1e8d55af79e95938b510

**Comment by bradzacher** [comment]
`satisfies` isn't a good alternative because it doesn't explicitly set the type.
Consider this example:
```ts
const f1 = (() => ('' as any));
//    ^? () => any

const f2 = (() => ('' as any)) satisfies () => string;
//    ^? () => any

const f3 = (() => ('' as any)) as () => string;
//    ^? () => string

const f4: () => string = () => ('' as any);
//    ^? () => string
```

`satisfies` only asserts that a value is assignable to a type - it doesn't change the type at all.
Which means that you can easily retain weak or incorrect contracts for your functions.

**Comment by Isti115** [comment]
Thanks for the comment @bradzacher! I know that `satisfies` doesn't actually apply the type to the expression, just validates the fact that it would fit the given annotation. It might as well be any subtype of it, so I agree with the possibility that it could lead to _"weak"_ annotations. As an extreme, people could even just write `... satisfies unknown` to suppress the warnings, but those who just want to get rid of it without providing useful information could just as easily disable the rule for the given line.

On the other hand, I know that you value the additional information provided through explicit annotations, as I fully agree with [one of your earlier comments](https://github.com/typescript-eslint/typescript-eslint/issues/1423#issuecomment-572901633) regarding the matter:
> _"I personally love explicit return types on everything for the reasons I stated (clear contracts, and easy to review)"_

I think that accepting `satisfies` would be a viable option for people who would like to provide additional information (and most likely would annotate with types strong enough to provide sufficient context for those reading the code later) without sacrificing on the correctness and convenience. Currently the two alternatives are either extracting the expression into a separate definition or using `as`, which makes it even easier to create incorrect typings.
